### ISSUE TITLE
Support `PCRE` `MARK` verb

### ISSUE THREAD

**Issue by Blacksmoke16** [crystal-lang/crystal#11320]
https://www.pcre.org/original/doc/html/pcrepattern.html and search for `Recording which path was taken`. Would need to add some additional bindings to `LibPCRE` and expose this value via `MatchData#mark`. Something like:

```cr
match = /X(*:A)Y|X(*:B)Z/.match("XZ").not_nil!
match.mark # => "B"
```

Related to https://github.com/crystal-lang/crystal/issues/8199, one con of this would be locking us into `PCRE`, given I'm not so sure other engines support it. However maybe https://github.com/crystal-lang/crystal/issues/3852 would fix it by optimizing the regex to be more performant?

**Comment by straight-shoota** [comment]
I'd call that a very specific use case. I'm not sure this is really necessary for a standard implementation.

Your router could surely just use custom libpcre bindings to get what you want for this particular use case.

In light of #11331, I would certainly not invest in this feature unless we're sure it will continue to work when we switch to a different engine.

**Comment by Blacksmoke16** [comment]
So now that we decided to go the PCRE2 route, this issue is essentially unblocked. Of which it can be implemented for both versions to remain backwards compatible. However the main question I have is how to best implement it given the engines have been abstracted out.

Probably be best to add it to the public API, but maybe add a note the underlying regex engine needs to support it? I.e. if we switch to something that doesn't it'll just always return `nil` or something. While in our case both PCRE versions can fill it in. Tho could also handle that down the line given the engines aren't really meant to be easily swapped out at the moment.

**Comment by straight-shoota** [comment]
I'm just a bit hesitant about adding support for such lesser-used regex features to the stdlib API. Your use case for routing looks good, but there don't seem to be many more (at least nobody else has asked for it yet). So this doesn't really feel like a stdlib feature to me.
Sure, implementing isn't hard. And I think we don't need to even support PCRE for a new feature. If you want to use it, update to PCRE2.
But still we'd be supporting something that's more of a niche feature, which means caring for maintenance. And as you mentioned, a potential problems for switching regex engines (which isn't really planned at this point and we should hopefully be good with PCRE2 for a very long time, so I don't worry too much about that).

Surely, you can also use a custom API for such specific features in your shard as you currently do. IMO that's not a bad choice for this. Perhaps it could be extracted to a generic shard for easy re-usability. There are other more specialized regex features that could fit there as well (for example partial matching). I guess it would be left to figure out if and how this could be integrated with the stdlib `Regex` class, though.
But I think I'd prefer outsourcing extended regex features into a shard over integrating them into stdlib.

**Comment by Blacksmoke16** [comment]
I guess my main thought is it just feels less than ideal to maintain essentially a duplicate of the code/bindings just so I can add a handful of lines. Then you need to make sure all the regexes that need it are using this custom implementation, can't really make use of regex literals, and any bug fixes/new features to stdlib Regex types need to be back ported.

Now that we're going for PCRE2, there is less code that would need to be duplicated, and I could prob monkey patch things in, but those core methods (like the match data constructor and such) are things that would have to be touched and kept up to date which is kind of annoying.

Maybe there is room for a cleaner way to add additional Regex features that doesn't involve duplicating/monkey patching the stdlib implementation at all.
